contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -1005,6 +1005,47 @@ fooRef:
 
 Although not always necessary to help a controller identify a resource type, “group” is included to avoid ambiguity when the resource exists in multiple groups. It also provides clarity to end users and enables copy-pasting of a reference without the referenced type changing due to a different controller handling the reference.
 
+##### Kind vs. Resource
+
+A common point of confusion in object references is whether to construct
+references with a "kind" or "resource" field. Historically most object
+references in Kubernetes have used "kind". This is not as precise as "resource".
+Although each combination of "group" and "resource" must be unique within
+Kubernetes, the same is not always true for "group" and "kind". It is possible
+for multiple resources to make use of the same "kind".
+
+Typically all objects in Kubernetes have a canonical primary resource - such as
+“pods” representing the way to create and delete resources of the “Pod” schema.
+While it is possible a resource schema cannot be directly created, such as a
+“Scale” object which is only used within the “scale” subresource of a number of
+workloads, most object references address the primary resource via its schema
+("kind").
+
+If an object reference may be used to reference resources of arbitrary types and
+the mapping between kind and resource could be ambiguous, "resource" should be
+used in the object reference.
+
+If implementations of an object reference will always have a clear way to map
```

Comment:
can you give an example of what a clear way would be? In particular, there's common use cases today that actually don't provide this guarantee. The big one to me is using RestMapper, and by extension any kind-based queries for controller-runtime: anyone can mount a new resource that re-uses someone else's kind, and effectively break this pattern.

In general I think this is a good discussion for follow-up work that needs to be done: we're going to tell people that you need to provide a direct resource-kind mapping, but then give implementors no facilities to make that easy (e.g. you can't insert direct resource-kind mappings into RestMapper, I believe).

But the gap in functionality aside, the risk should be clarified. restmapper puts you at risk, and hand-wiring is the only safe option.

Reply:
Updated, thanks!